File: app/models/supplier.rb
class Supplier < ApplicationRecord
    has_many :invoices
    has_many :invoice_transactions, through: :invoices

    def self.transaction_counts
        left_joins(:invoices, :invoice_transactions)
        .group(:id)
        .order(created_at: :desc)
        .select(:id, :name, 'count(suppliers.id) AS orders_count,
            sum(invoice_transactions.amount) AS amount,
            sum(CASE WHEN invoices.order_received_date IS NULL THEN invoice_transactions.amount ELSE 0 END) as pending_amount,
            sum(CASE WHEN invoices.order_received_date IS NOT NULL THEN invoice_transactions.amount ELSE 0 END) as received_amount')

    end

    def add_invoice(order_date, order_number, invoiced, billed)
        Invoice.new(:supplier => self,
            :order_date => Time.parse(order_date),
            :order_number => order_number,
            :invoiced_amount => invoiced,
            :billed_amount => billed)
    end
end

File: frontend/src/app/supplier/supplier.ts
import { GetSuppliersResponse } from './supplier.data.response';

export class Supplier {
    private _id: number;

    public get id() {
        return this._id;
    }

    public name: string;
    public amount: number;
    public pending_amount: number;

    constructor(response?: GetSuppliersResponse) {
        if (response) {
            this._id = response.id;
            this.name = response.name;
            this.amount = response.amount;
            this.pending_amount = response.pending_amount;
        }
    }

    public toString() {
        return this.name;
    }
}
